src/v/pandaproxy/schema_registry/handlers.cc

```diff
@@ -259,38 +251,42 @@ delete_subject_version(server::request_t rq, server::reply_t rp) {
         .value_or(permanent_delete::no)};
     rq.req.reset();
 
+    // Must see latest data to know whether what we're deleting is the last
+    // version
+    co_await rq.service().writer().read_sync();
+
     auto version = invalid_schema_version;
+    bool final_version = false;
     if (ver == "latest") {
+        if (permanent == permanent_delete::yes) {
+            // Permanent deletes on 'latest' not allowed.  They are
+            // transformed into soft deletes
+            permanent = permanent_delete::no;
+        }
+
+        // Requests for 'latest' mean the latest which is not marked deleted
         auto versions = co_await rq.service().schema_store().get_versions(
-          sub, include_deleted::yes);
+          sub, include_deleted::no);
         if (versions.empty()) {
             throw as_exception(not_found(sub, version));
         }
         version = versions.back();
     } else {
+        auto versions = co_await rq.service().schema_store().get_versions(
+          sub, include_deleted::yes);
+        if (versions.size() == 1) {
+            final_version = true;
+        }
         version = parse::from_chars<schema_version>{}(ver).value();
     }
 
-    auto d_res = co_await rq.service().schema_store().delete_subject_version(
-      sub, version, permanent, include_deleted::no);
-
-    if (d_res) {
-        std::optional<model::record_batch> batch;
-        if (permanent) {
-            batch.emplace(
-              make_delete_subject_version_permanently_batch(sub, version));
-        } else {
-            auto s_res
-              = co_await rq.service().schema_store().get_subject_schema(
-                sub, version, include_deleted::yes);
-            batch.emplace(make_delete_subject_version_batch(std::move(s_res)));
-        }
-        auto res = co_await rq.service().client().local().produce_record_batch(
-          model::schema_registry_internal_tp, std::move(batch).value());
-
-        if (res.error_code != kafka::error_code::none) {
-            throw kafka::exception(res.error_code, *res.error_message);
-        }
+    // A permanent deletion emits tombstones for prior schema_key messages
+    if (permanent) {
+        co_await rq.service().writer().delete_subject_permanent(
+          sub, final_version ? std::nullopt : std::make_optional(version));
```

Comment:
This seems to have broken some logic, I'm a state where deleting a version is telling me I haven't soft-deleted the subject

Reply:
@jcsp ?

Reply:
This is fixed now.  There's a new test called `test_mixed_deletes` based on the series of curl operations that found it.

I fixed a couple of other issues in hard deletes exposed in that test's checks on subjects appearing or not in the list, and some log noise on tombstones that I noticed at the time.

Those are squashed in on this PR, but visible as individual commits on https://github.com/jcsp/redpanda/commits/schema-registry-ha-fixes for anyone wanting to look at those changes to the logic (it's mostly additions of cases where we should be updating things we weren't)